Layout.tsx を修正し、ログイン・サインアップ画面用のヘッダーを作成

toshinori-m opened the pull request:
### 変更内容
1. LoginHeader.tsx の追加
- react-router-dom の useNavigate を使用し、トップページ (/) へのリダイレクト機能を持つ OpenHeader コンポーネントを作成。
- stay_connect のボタンを追加し、クリックでトップページへ遷移。
2. Layout.tsx の修正
- useLocation を導入し、現在のパスに応じて表示するヘッダー (OpenHeader または LoginHeader) を動的に変更。
  - / の場合は OpenHeader を表示。
  - /signup の場合は LoginHeader を表示。
  - それ以外のパスでは OpenHeader をデフォルトとして設定。
### 動作確認方法
- `http://localhost:5173/signup` 、`http://localhost:5173 にアクセスへヘッダーが切り替わることを確認。



close https://github.com/toshinori-m/stay_connect/issues/159

toshinori-m committed:
ログイン・サインアップ画面用のヘッダーを追加

- LoginHeader.tsx を新規作成し、ログイン/サインアップ/登録ページ用のヘッダーを実装
- Layout.tsx を修正し、現在のパス (login) に応じて適切なヘッダーを表示

hirosi1900day commented on frontend-react/src/components/layout/Layout.tsx:
```diff
@@ -1,10 +1,26 @@
+      return <LoginHeader />
+    } else {
+      return <OpenHeader />
+    }
```
パスで管理するよりもログインしているかしてないかでheaderを出し分ける方が管理しやすいかなと思いましたがいかがでしょうか？？
ここをパスで制御してしまうと、signupなどのパスをregisterなどに変更したいときなどに修正漏れが発生してバグの原因になりそうな気がします(運用した時に楽になる実装が好ましいと思います)

toshinori-m replied:
"/"および"/signup"はログイン前の画面なのでパスで変更するようにしていました。

今回も同じようなコードになっているかもしれませんが、ログイン状態を優先する形にしました

hirosi1900day replied:
あ〜
若干内容をもしかすると勘違いされているかもです(書き方が悪かったです)

できればパスを使わずに未ログインとログイン状態のヘッダー２種類で管理し、ログインしているかいないかだけで出しわけするイメージで指摘しておりました。

それが無理そうで、もしパスを使ってlayoutを切り替えたいのであればパス制御の責務を担うreact routerの部分で切り替えるといいかもしれないですね！(別issueでも大丈夫です)
https://ar-aca.tech/posts/react-router-layout/

ex:
```tsx
<Routes>
  <Route path="/" element={<Layout />}>
    <Route index element={<Home />} />
    <Route path="about" element={<About />} />
  </Route><Route path="/dashboard" element={<DashboardLayout />}>
    <Route index element={<DashboardHome />} />
  </Route>
</Routes>
```

toshinori-m replied:
下のようにコードを修正しようとしたところ「名前空間 'JSX' が見つかりません。」とエラーが出て解決できないので教えてください
`const headerMap: Record<string, JSX.Element>`

試したこと:
```
tsconfig.jsonに下を追加しコンテナ再起動
{
  "compilerOptions": {
    "jsxImportSource": "react",
    "types": ["react", "react-dom"],
  }
}
```
そして、下の記事を見ていますが理解が難しいです
https://qiita.com/uhyo/items/adf6cb83333a25097f25

よろしくお願いします

hirosi1900day replied:
>責務を担うreact routerの部分で切り替えるといいかもしれないですね！(別issueでも大丈夫です)
https://ar-aca.tech/posts/react-router-layout/

こちらの方針の方が良さそうな気がしてます！

hirosi1900day replied:
指摘の意図としては
複数の箇所でパスによって処理を切り替えているとパスを変更したくなった際に修正漏れが発生する可能性があるので
パスによって処理を切り替える処理は極力減らしたい or 一箇所に集中させたい(routerに集約したい)

(普通のwebアプリだと開発するエンジニアが入れ替わるので複数の箇所に処理が散らばっていると認知負荷が上がり、修正漏れなどが発生する可能性があるためそう言った実装をできるだけ減らしたいという意図)

とはいえそれが難しそうであればそのままの実装でも良いかなと思います！(そこに時間をかけすぎるのもアレなので)

toshinori-m replied:
ありがとうございます！
今回はできているように感じます

toshinori-m committed:
ヘッダーの制御をパス依存からログイン状態優先に変更

toshinori-m commented:
本来は別の issue で対応すべき作業ですが、ヘッダーの表示切り替えのタイミングが良かったので、下の理由で SignupPage.tsx を修正しました。

- ユーザーがログインしたまま PC を離れ、タブを閉じた場合、再度 SignupPage.tsx を開くとログイン状態が維持されてしまう問題があったため

toshinori-m committed:
React Router でヘッダーを適用するように変更

- Layout.tsx を削除し、ルートごとに適切な Layout を適用
- OpenLayout, SignupLayoutを router/index.tsx に追加

hirosi1900day reviewed (approved):
LGTM ！
